packages/cli/src/commands/update.js

```diff
@@ -0,0 +1,251 @@
+      console.log(
+        'An update command is already running, please try again in a few seconds.'
+      )
+    }
+    return
+  }
+
+  let upgradeAvailable = false
+  const updateTasks = new Listr([], { collapse: false })
+
+  if (silent) {
+    updateTasks.setRenderer('silent')
+  }
+
+  updateTasks.add({
+    title: 'Checking if a newer RedwoodJS version is available',
+    task: async (ctx, task) => {
+      const versionStatus = await getUpdateVersionStatus()
+      upgradeAvailable = versionStatus.upgradeAvailable
+
+      if (upgradeAvailable) {
+        createUpgradeVersionsFile(versionStatus)
+        if (!silent) {
+          // TODO: prompt if they want to upgrade now?
+          // TODO: add on exit hook to run `yarn rw upgrade`
+          task.title = 'New upgrade is available'
+        }
+      } else {
+        removeUpgradeVersionsFile()
+        task.title = 'No upgrade is available'
+      }
+
+      resetUpdateTimer()
+    },
+  })
+
+  try {
+    createUpdateLockFile()
+    await updateTasks.run()
+    if (upgradeAvailable && !silent) {
+      console.log(upgradeAvailableMessage())
+    }
+  } catch (e) {
+    errorTelemetry(process.argv, e.message)
+    console.error(c.error(e.message))
+    process.exit(e?.exitCode || 1)
+  } finally {
+    removeUpdateLockFile()
+  }
+}
+
+// Locking
+
+function getUpdateLockFilePath() {
+  return path.join(getPaths().base || '/tmp', '.redwood', 'update', 'lock')
+}
+function createUpdateLockFile() {
+  try {
+    fs.writeFileSync(getUpdateLockFilePath(), '')
+  } catch (error) {
+    if (error.code === 'ENOENT') {
+      fs.mkdirSync(path.dirname(getUpdateLockFilePath()))
+      fs.writeFileSync(getUpdateLockFilePath(), '')
+    }
+  }
+}
+function removeUpdateLockFile() {
+  fs.unlinkSync(getUpdateLockFilePath())
+}
+function isUpdateLocked() {
+  return fs.existsSync(getUpdateLockFilePath())
+}
+
+// Flag file
+
+function getUpgradeVersionsFilePath() {
+  return path.join(
+    getPaths().base || '/tmp',
+    '.redwood',
+    'update',
+    'versions.json'
+  )
+}
+
+function createUpgradeVersionsFile(versionStatus) {
+  fs.writeFileSync(getUpgradeVersionsFilePath(), JSON.stringify(versionStatus))
+}
+
+function removeUpgradeVersionsFile() {
+  try {
+    fs.unlinkSync(getUpgradeVersionsFilePath())
+  } catch (error) {
+    if (error.code !== 'ENOENT') {
+      throw new Error('\nCould not delete update versions file\n')
+    }
+  }
+}
+
+export function isUpgradeAvailable() {
+  return fs.existsSync(getUpgradeVersionsFilePath())
+}
+
+function readUpgradeVersionsFile() {
+  return JSON.parse(fs.readFileSync(getUpgradeVersionsFilePath()))
+}
+
+// TODO: Confirm this is a sensible way to check the current version
+async function getUpdateVersionStatus() {
+  const packageJson = require(path.join(getPaths().base, 'package.json'))
+  let localVersion = packageJson.devDependencies['@redwoodjs/core']
+  while (!(localVersion.charAt(0) >= '0' && localVersion.charAt(0) <= '9')) {
```

Comment:
In what scenario is this code needed? Could you please add some code comments, describing what's going on and showing some example input? Thanks :) 

And this would be a great method to add some unit tests for

Reply:
Okay I'll add some comments to hopefully make the process clearer.

I'm also very eager to add tests for the whole command but this is an area in which I have little experience. I plan to take an hour or two and learn more about testing. After that I'll give adding some tests a go.